mempool/clist_mempool.go

```diff
@@ -701,38 +774,45 @@ func (mem *CListMempool) recheckTxs() {
 		mem.logger.Error("Not all txs were rechecked", "not-rechecked", n)
 	}
-	mem.logger.Debug("Done rechecking txs", "height", mem.height.Load(), "num-txs", mem.Size())
+
+	mem.logger.Debug("Done rechecking", "height", mem.height.Load(), "num-txs", mem.Size())
 }
 
-// The cursor and end pointers define a dynamic list of transactions that could be rechecked. The
-// end pointer is fixed. When a recheck response for a transaction is received, cursor will point to
-// the entry in the mempool corresponding to that transaction, thus narrowing the list. Transactions
-// corresponding to entries between the old and current positions of cursor will be ignored for
-// rechecking. This is to guarantee that recheck responses are processed in the same sequential
-// order as they appear in the mempool.
+// When a recheck response for a transaction is received, cursor will point to
+// the entry in the mempool corresponding to that transaction, advancing the
+// cursor, thus narrowing the list of transactions to recheck. In case there are
+// entries between the previous and the current positions of cursor, they will
+// be ignored for rechecking. This is to guarantee that recheck responses are
+// processed in the same sequential order as they appear in the mempool.
 type recheck struct {
-	cursor        *clist.CElement // next expected recheck response
-	end           *clist.CElement // last entry in the mempool to recheck
-	doneCh        chan struct{}   // to signal that rechecking has finished successfully (for async app connections)
-	numPendingTxs atomic.Int32    // number of transactions still pending to recheck
-	isRechecking  atomic.Bool     // true iff the rechecking process has begun and is not yet finished
-	recheckFull   atomic.Bool     // whether rechecking TXs cannot be completed before a new block is decided
+	iter          *NonBlockingIterator
+	cursor        Entry         // next expected recheck response
+	doneCh        chan struct{} // to signal that rechecking has finished successfully (for async app connections)
+	numPendingTxs atomic.Int32  // number of transactions still pending to recheck
+	isRechecking  atomic.Bool   // true iff the rechecking process has begun and is not yet finished
+	recheckFull   atomic.Bool   // whether rechecking TXs cannot be completed before a new block is decided
 }
 
-func newRecheck() *recheck {
-	return &recheck{
-		doneCh: make(chan struct{}, 1),
-	}
+func newRecheck(iter *NonBlockingIterator) *recheck {
+	r := recheck{}
+	r.iter = iter
+	return &r
 }
 
-func (rc *recheck) init(first, last *clist.CElement) {
+func (rc *recheck) init(mp *CListMempool) {
 	if !rc.done() {
 		panic("Having more than one rechecking process at a time is not possible.")
 	}
-	rc.cursor = first
-	rc.end = last
 	rc.numPendingTxs.Store(0)
+	rc.iter = NewNonBlockingIterator(mp)
+
+	rc.cursor = rc.iter.Next()
+	if rc.cursor == nil {
+		return
+	}
+	rc.doneCh = make(chan struct{})
 	rc.isRechecking.Store(true)
```

Comment:
But then, if the mempool is empty

* we won't setup the channel... but later in the code we receive from than channel, will that work?
* about `isRechecking` I'm less sure, but just in case, shouldn't we make sure we set it to false here?

Reply:
`isDone` checks whether `isRechecking` is false. 
About the channel you are right, we might want to not return here early because even with setting the other stuff, if the mempool is empty it will be set downstream.

Reply:
Yes, for the channel, we are calling `doneRechecking()` inside `recheckTxs`, whether the mempool is empty or not. So, if the mempool is empty, we probably need to create it, and then close it right away. Otherwise, we are using the channel from the previous "incarnation" of rechecktx. Maybe we can trace the code and convince ourselves that the channel exists and is closed anyway, but what about the "first incarnation"?

Reply:
Hmm, actually we don't close the channel if reCheckTx times out 🤦

Reply:
Let me propose something... bear with me

Reply:
How about this?
```
diff --git a/mempool/clist_mempool.go b/mempool/clist_mempool.go
index 7b97ce52f1..64b551d6a4 100644
--- a/mempool/clist_mempool.go
+++ b/mempool/clist_mempool.go
@@ -807,10 +807,11 @@ func (rc *recheck) init(mp *CListMempool) {
 	rc.iter = NewNonBlockingIterator(mp)
 
 	rc.cursor = rc.iter.Next()
+	rc.doneCh = make(chan struct{})
 	if rc.cursor == nil {
+		rc.setDone()
 		return
 	}
-	rc.doneCh = make(chan struct{})
 	rc.isRechecking.Store(true)
 	rc.recheckFull.Store(false)
 }
@@ -826,6 +827,7 @@ func (rc *recheck) setDone() {
 	rc.cursor = nil
 	rc.recheckFull.Store(false)
 	rc.isRechecking.Store(false)
+	close(rc.doneCh) // notify channel that recheck has finished
 }
 
 // findNextEntryMatching searches for the next transaction matching the given transaction, which
@@ -849,7 +851,6 @@ func (rc *recheck) findNextEntryMatching(tx *types.Tx) (found bool) {
 
 	if rc.cursor == nil { // reached end of list
 		rc.setDone()
-		close(rc.doneCh) // notify channel that recheck has finished
 	}
 	return found
 }
```

Reply:
I think this last proposal works well.

Reply:
Still, `recheck` has many atomic variables when actually they should all be guarded by one mutex. We haven't hit a race condition yet.

Reply:
Added the code Sergio wrote.

Reply:
@jmalicevic thanks!

@hvanz I agree. Maybe we can get rid of all the atomic vars, and protect them with one single mutex. And also, see if we can have one mutex instead of two. But all that should be done in a further PR